archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,874 @@
+-/
+lemma monotone_convergence (bn : ℕ → ℝ) (h_sd : ∀ (n m : ℕ), n ≤ m → bn m ≤ bn n)
+  (h_bounded : (lower_bounds (set.range bn)).nonempty) : ∃ (m : ℝ), tendsto bn at_top (𝓝 m) :=
+begin
+ use Inf (set.range bn),
+ exact tendsto_at_top_is_glb h_sd (real.is_glb_Inf (set.range bn)
+   (set.range_nonempty bn) h_bounded),
+end
+
+/-- The sequence `an` is positive for `n > 0`  -/
+lemma an'_pos : ∀ (n : ℕ), 0 < an n.succ :=
+ (λ n, div_pos (cast_pos.mpr (factorial_pos n.succ))
+    (mul_pos ((real.sqrt_pos).mpr (mul_pos two_pos (cast_pos.mpr (succ_pos n))))
+    (pow_pos (div_pos (cast_pos.mpr (succ_pos n)) (exp_pos 1)) n.succ)))
+
+/--
+The sequence `an` has the explicit lower bound exp (3/4 - 1/2 * log 2)
+-/
+lemma an'_bounded_by_pos_constant : ∀ (n : ℕ), exp (3 / (4 : ℝ) - 1 / 2 * log 2) ≤ an n.succ :=
+begin
+  intro n,
+  rw ← (le_log_iff_exp_le (an'_pos n)),
+  exact bn_bounded_by_constant n,
+end
+
+/-- The sequence `an` is monotone decreasing -/
+lemma an'_antitone : ∀ (n m : ℕ), n ≤ m → an m.succ ≤ an n.succ :=
+  (λ n, λ m, λ h, (log_le_log (an'_pos m) (an'_pos n)).mp (bn_antitone n m h))
+
+/-- The sequence `an` has a lower bound -/
+lemma an'_has_lower_bound : (lower_bounds (set.range (λ (k : ℕ), an k.succ))).nonempty :=
+begin
+   use exp (3 / (4 : ℝ) - 1 / 2 * log 2),
+   intros,
+   rw lower_bounds,
+   simp only [set.mem_range, forall_exists_index, forall_apply_eq_imp_iff', set.mem_set_of_eq],
+   exact an'_bounded_by_pos_constant,
+end
+
+/-- The sequence `an` converges to a real number -/
+lemma an_has_limit_a : ∃ (a : ℝ), tendsto (λ (n : ℕ), an n) at_top (𝓝 a) :=
+begin
+  have ha := monotone_convergence (λ (k : ℕ), an k.succ) an'_antitone an'_has_lower_bound,
+  cases ha with x hx,
+  rw ← tendsto_succ an x at hx,
+  use x,
+  exact hx,
+end
+
+/-- The limit `a` of the sequence `an` satisfies `0 < a` -/
+lemma an_has_pos_limit_a : ∃ (a : ℝ), 0 < a ∧ tendsto (λ (n : ℕ), an n) at_top (𝓝 a) :=
+begin
+  have h := an_has_limit_a,
+  cases h with a ha,
+  use a,
+  split,
+  { let an' : ℕ → ℝ := λ n, an n.succ,
+    rw tendsto_succ an a at ha,
+    have e_lower_bound : exp (3 / (4 : ℝ) - 1 / 2 * log 2) ∈ lower_bounds (set.range an') :=
+    begin
+      intros x hx,
+      rw [set.mem_range] at hx,
+      cases hx with k hk,
+      rw ← hk,
+      exact an'_bounded_by_pos_constant k,
+    end,
+    exact gt_of_ge_of_gt ((le_is_glb_iff (is_glb_of_tendsto_at_top an'_antitone ha)).mpr
+      e_lower_bound) (3 / 4 - 1 / 2 * log 2).exp_pos },
+  { exact ha },
+end
+
+/-!
+ ### Part 2
+ https://proofwiki.org/wiki/Stirling%27s_Formula#Part_2
+-/
+
+/--
+Define `wallis_inside_prod n` as `2 * n / (2 * n - 1) * 2 * n/(2 * n + 1)`.
+This is the term appearing inside the Wallis product
+-/
+noncomputable def wallis_inside_prod (n : ℕ) : ℝ :=
+  (((2 : ℝ) * n) / (2 * n - 1)) * ((2 * n) / (2 * n + 1))
+
+/--
+We can shift the indexing in a product.
+ -/
+lemma aux1 (k : ℕ) : ∏ i in range k, (wallis_inside_prod (1 + i)) =
```

Comment:
Is this not a much more general theorem?

Reply:
Yes, sure, it doesn't depend at all on the specific function. Perhaps this is something to include somewhere in the big_operators part of mathlib. (I hope I didn't miss it there...)

Reply:
It's basically `prod_Ico_add` and `range_eq_Ico` put together (which is how it was proven). Here, one has to fiddle around a bit to get the `n+1` changed to `succ n` inside the product. Since it was used exactly once, I deleted the lemma and inserted the proof at the relevant location.